AuditEventProcessor.cpp

```diff
@@ -122,13 +129,159 @@ void AuditEventProcessor::static_callback(void *au, dummy_enum_t cb_event_type,
     processor->callback(au);
 }
 
+bool AuditEventProcessor::process_execve()
+{
+    static const std::unordered_set<std::string> execve_fields = { "arch", "syscall", "success", "exit", "items", "ppid", "pid", "auid", "uid", "gid", "euid", "suid", "fsuid", "egid", "sgid", "fsgid", "tty", "ses", "comm", "exe", "key", "name", "inode", "dev", "mode", "ouid", "ogid", "rdev", "nametype", "cwd", "cmdline" };
+
+    if (_num_records < 5 || auparse_get_type(_state) != AUDIT_SYSCALL) {
+        return false;
+    }
+
+    const char *syscall = auparse_find_field(_state, "syscall");
+    auparse_first_field(_state);
+
+    if (syscall == nullptr || atoi(syscall) != SYS_execve) {
+        return false;
+    }
+
+    auto ret = _builder->BeginEvent(_current_event_sec, _current_event_msec, _current_event_serial, 1);
+    if (ret != 1) {
+        if (ret == Queue::CLOSED) {
+            throw std::runtime_error("Queue closed");
+        }
+        return false;
+    }
+
+    ret = _builder->BeginRecord(PROCESS_CREATE_RECORD_TYPE, PROCESS_CREATE_RECORD_NAME, "", execve_fields.size());
+    if (ret != 1) {
+        if (ret == Queue::CLOSED) {
+            throw std::runtime_error("Queue closed");
+        }
+        cancel_event();
+        return false;
+    }
+
+    do {
+        int record_type = auparse_get_type(_state);
+
+        switch (record_type) {
+            case AUDIT_SYSCALL: {
+                do {
+                    const char* field = auparse_get_field_name(_state);
+                    if (field != nullptr && execve_fields.count(std::string(field)) && !process_field(field)) {
+                        cancel_event();
+                        return false;
+                    }
+                } while (auparse_next_field(_state) == 1);
+                break;
+            }
+            case AUDIT_EXECVE: {
+                std::string commandline = "";
```

Comment:
Please move this to a member variable. See OMSEventWriter::process_field() and _raw_name for an example. This will help avoid the malloc/realloc/free that would otherwise be necessary for each event.